### PR TITLE
fix(prometheus_scrape source): Don't crash on out-of-order histograms 

### LINKED ISSUE #7999: Failure to send histograms scraped from prometheus_scrape through prometheus_remote_write sink to prometheus_remote_write source
### Vector Version

```
vector 0.14.0 (x86_64-unknown-linux-gnu 5f3a319 2021-06-03)
```

### Vector Configuration File

Vector A:

```toml
[sources.prometheus_metrics]
  type                          = "prometheus_scrape"
  endpoints                     = ["http://127.0.0.1:8000/metrics"]
  scrape_interval_secs          = 5


[transforms.prometheus_tagged]
  type                          = "remap"
  inputs                        = ["prometheus_metrics"]
  source                        = '''
  .tags.stack_id = "GitLabInfra"
  .tags.account_id = "630000000000"
  .tags.instance_id = "i-0ffa1853550a55cd8"
  '''

[sinks.sink_ingest]
  type                          = "prometheus_remote_write"
  inputs                        = ["prometheus_tagged"]
  default_namespace             = "node"
  endpoint                      = "http://localhost:5623"
```

Vector B:

```toml
[sources.source_prometheus_writer]
type                      = "prometheus_remote_write"
address                   = "0.0.0.0:5623"

[sinks.console]
type = "console"
inputs = ["source_prometheus_writer"]
encoding.codec = "json"
```

### Debug Output

Vector A:

```
Jun 23 01:01:05.701  INFO vector::app: Log level is enabled. level="vector=info,codec=info,vrl=info,file_source=info,tower_limit=trace,rdkafka=info"
Jun 23 01:01:05.701  INFO vector::sources::host_metrics: PROCFS_ROOT is unset. Using default /proc for procfs root.
Jun 23 01:01:05.701  INFO vector::sources::host_metrics: SYSFS_ROOT is unset. Using default /sys for sysfs root.
Jun 23 01:01:05.701  INFO vector::app: Loading configs. path=[("/tmp/in.toml", None)]
Jun 23 01:01:05.710  INFO vector::topology: Running healthchecks.
Jun 23 01:01:05.710  INFO vector::topology: Starting source. name="prometheus_metrics"
Jun 23 01:01:05.710  INFO vector::topology: Starting transform. name="prometheus_tagged"
Jun 23 01:01:05.710  INFO vector::topology: Starting sink. name="sink_ingest"
Jun 23 01:01:05.710  INFO vector: Vector has started. version="0.14.0" arch="x86_64" build_id="5f3a319 2021-06-03"
Jun 23 01:01:05.711  INFO vector::app: API is disabled, enable by setting `api.enabled` to `true` and use commands like `vector top`.
Jun 23 01:01:05.711 ERROR vector::topology::builder: msg="Healthcheck: Failed Reason." error=Unexpected status: 405 Method Not Allowed component_kind="sink" component_type="prometheus_remote_write" component_name="sink_ingest"
Jun 23 01:01:05.757 ERROR sink{component_kind="sink" component_name=sink_ingest component_type=prometheus_remote_write}:request{request_id=0}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 400 Bad Request"
Jun 23 01:01:05.757 ERROR sink{component_kind="sink" component_name=sink_ingest component_type=prometheus_remote_write}:request{request_id=0}: vector::sinks::util::sink: Response failed. response=Response { status: 400, version: HTTP/1.1, headers: {"content-type": "application/json", "content-length": "115", "date": "Wed, 23 Jun 2021 01:01:05 GMT"}, body: b"{\"code\":400,\"message\":\"Could not decode write request: expected value in range [0, 4294967295], found: 4294967296\"}" }
Jun 23 01:01:20.735 ERROR sink{component_kind="sink" component_name=sink_ingest component_type=prometheus_remote_write}:request{request_id=2}: vector::sinks::util::retries: Not retriable; dropping the request. reason="response status: 400 Bad Request"
Jun 23 01:01:20.736 ERROR sink{component_kind="sink" component_name=sink_ingest component_type=prometheus_remote_write}:request{request_id=2}: vector::sinks::util::sink: Response failed. response=Response { status: 400, version: HTTP/1.1, headers: {"content-type": "application/json", "content-length": "115", "date": "Wed, 23 Jun 2021 01:01:20 GMT"}, body: b"{\"code\":400,\"message\":\"Could not decode write request: expected value in range [0, 4294967295], found: 4294967331\"}" }
```

### Expected Behavior

The histogram should be scraped by Vector A and sent through to Vector B to be printed out in the console.

### Actual Behavior

The histogram is not sent successfully and an error is output.

### Example Data

I used Python's test HTTP server to serve up:

```
# HELP duration_seconds Multiprocess metric
# TYPE duration_seconds histogram
duration_seconds_bucket{le="+Inf"} 2
duration_seconds_bucket{le="0.001"} 0
duration_seconds_bucket{le="0.01"} 2
duration_seconds_bucket{le="0.1"} 2
duration_seconds_bucket{le="1"} 2
duration_seconds_count 2
duration_seconds_sum 0.010685475077480078
```

As prometheus scrape output.

### References

- #7342

### LINKED ISSUE #8012: Parsing invalid Prometheus text metrics can crash Vector
When the `prometheus_scrape` source is presented with histogram metrics with the fields out of order, the source thread can panic with the error `thread 'tokio-runtime-worker' panicked at 'attempt to subtract with overflow', src/sources/prometheus/parser.rs:75:25`

Sample data file producing the error:
```
# HELP duration_seconds Multiprocess metric
# TYPE duration_seconds histogram
duration_seconds_bucket{le="0.001"} 0
duration_seconds_bucket{le="0.01"} 0
duration_seconds_bucket{le="0.1"} 0
duration_seconds_bucket{le="1"} 2
duration_seconds_bucket{le="+Inf"} 2
duration_seconds_count 2
duration_seconds_sum 0.010685475077480078
```

The fix should be to enforce the buckets are in the right order before doing the subtractions causing the overflow.

### PR DESCRIPTION
**Pull request by bruceg**
Closes #8012

### CODE REVIEW

File: lib/prometheus-parser/src/lib.rs [hunk 1]
```diff
@@ -79,7 +79,7 @@ pub struct SummaryMetric {
     pub count: u32,
 }
 
-#[derive(Debug, Default, PartialEq)]
+#[derive(Debug, Default, PartialEq, PartialOrd)]
```
Comment (jszwedko) [inline lib/prometheus-parser/src/lib.rs:82]: Is there a reason we don't just derive or implement `Ord`?

**Review by jszwedko — APPROVED** [review]
Thanks for this quick fix @bruceg ! Just one question about whether we should just implement `Ord` and then use `buckets.sort()`, but otherwise this looks good. I agree with the addition of `saturating_sub` too.

File: lib/prometheus-parser/src/lib.rs (hunk 1 above)
Comment (bruceg) [reply lib/prometheus-parser/src/lib.rs:82]: The "total order" predicate required for `Ord` semantics on floating point values is not implemented in hardware, so actually implementing this in Rust is messy at best. See https://stackoverflow.com/questions/26489701/why-does-rust-not-implement-total-ordering-via-the-ord-trait-for-f64-and-f32

Comment (bruceg) [reply lib/prometheus-parser/src/lib.rs:82]: FTR I created this based on the suggestion in the above link, which passes the trivial test, but I'm not confident it's actually correct for negatives and such. What's in this patch is much simpler.
```rust
impl Eq for HistogramBucket {}

impl Ord for HistogramBucket {
    fn cmp(&self, other: &Self) -> Ordering {
        self.bucket
            .is_sign_positive()
            .cmp(&other.bucket.is_sign_positive())
            .then_with(|| {
                (self.bucket.to_bits() | 0x7fffffffffffffffu64)
                    .cmp(&(other.bucket.to_bits() & 0x7fffffffffffffffu64))
            })
            .then_with(|| self.count.cmp(&other.count))
    }
}
```

Comment (jszwedko) [reply lib/prometheus-parser/src/lib.rs:82]: Aha, yes, I see. I neglected the bucket was an `f64`. In that case, I think what you have is good.

**Pull request merged**